Build failure: firewalk

### Nixpkgs version

- Unstable (26.05)

### Steps to reproduce

`nix-shell -p firewalk`

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/316554629

### Relevant log output

```console
gcc -DHAVE_CONFIG_H -I. -I. -I../include     -g -O2 -Wall -c init.c
init.c: In function 'fw_init_context':
init.c:62:27: error: passing argument 2 of 'catch_sig' from incompatible pointer type [-Wincompatible-pointer-types]
   62 |     if (catch_sig(SIGINT, catch_sigint) == -1)
      |                           ^~~~~~~~~~~~
      |                           |
      |                           void (*)(int)
In file included from init.c:38:
../include/firewalk.h:251:5: note: expected 'void (*)(void)' but argument is of type 'void (*)(int)'
  251 |     void (*)()                      /* new signal handler */
      |     ^~~~~~~~~~
../include/firewalk.h:256:1: note: 'catch_sigint' declared here
  256 | catch_sigint(
      | ^~~~~~~~~~~~
init.c: In function 'fw_init_net':
init.c:92:23: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
   92 |         (*fp)->device = libnet_getdevice((*fp)->l);
      |                       ^
make[1]: *** [Makefile:164: init.o] Error 1
make[1]: Leaving directory '/build/firewalk-upstream-5.0/src'
make: *** [Makefile:160: all-recursive] Error 1
error: Cannot build '/nix/store/vd6lj2h11v7vcjyws3gsaxla5ijg15sm-firewalk-5.0.drv'.
       Reason: builder failed with exit code 2.
       Output paths:
         /nix/store/4cg8wxz3rgscw7n8rx20lal39kjr8g95-firewalk-5.0
       Last 25 log lines:
       > Running phase: buildPhase
       > build flags: SHELL=/nix/store/lw117lsr8d585xs63kx5k233impyrq7q-bash-5.3p3/bin/bash
       > Making all in src
       > make[1]: Entering directory '/build/firewalk-upstream-5.0/src'
       > gcc -DHAVE_CONFIG_H -I. -I. -I../include     -g -O2 -Wall -c init.c
       > init.c: In function 'fw_init_context':
       > init.c:62:27: error: passing argument 2 of 'catch_sig' from incompatible pointer type [-Wincompatible-pointer-types]
       >    62 |     if (catch_sig(SIGINT, catch_sigint) == -1)
       >       |                           ^~~~~~~~~~~~
       >       |                           |
       >       |                           void (*)(int)
       > In file included from init.c:38:
       > ../include/firewalk.h:251:5: note: expected 'void (*)(void)' but argument is of type 'void (*)(int)'
       >   251 |     void (*)()                      /* new signal handler */
       >       |     ^~~~~~~~~~
       > ../include/firewalk.h:256:1: note: 'catch_sigint' declared here
       >   256 | catch_sigint(
       >       | ^~~~~~~~~~~~
       > init.c: In function 'fw_init_net':
       > init.c:92:23: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
       >    92 |         (*fp)->device = libnet_getdevice((*fp)->l);
       >       |                       ^
       > make[1]: *** [Makefile:164: init.o] Error 1
       > make[1]: Leaving directory '/build/firewalk-upstream-5.0/src'
       > make: *** [Makefile:160: all-recursive] Error 1
       For full logs, run:
         nix log /nix/store/vd6lj2h11v7vcjyws3gsaxla5ijg15sm-firewalk-5.0.drv
```

### Additional context

Firewalk doesn't seem to have had any changes in a long time, seems related to recent changes to libnet, which also seemed to break some other packages? Haven't tried building without those changes to double-check.

Recent changes to libnet:
- https://github.com/NixOS/nixpkgs/pull/467210
- https://github.com/NixOS/nixpkgs/pull/469339

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.61, NixOS, 26.05 (Yarara), 26.05.20260116.e4bae1b`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.31.2+2`
 - nixpkgs: `/nix/store/kfcxqcxb9hcq6x33sg4cmwakbb1ifwg9-source`

### Notify maintainers

@Tochiaha 

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+label%3A%220.kind%3A+build+failure%22).
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/fi/firewalk/package.nix
+++ b/pkgs/by-name/fi/firewalk/package.nix
@@ -2,6 +2,7 @@
   lib,
   stdenv,
   fetchurl,
+  fetchpatch,
   libnet,
   libpcap,
   libdnet,
@@ -15,6 +16,13 @@ stdenv.mkDerivation (finalAttrs: {
     url = "https://salsa.debian.org/pkg-security-team/firewalk/-/archive/upstream/${finalAttrs.version}/firewalk-upstream-${finalAttrs.version}.tar.gz";
     hash = "sha256-f0sHzcH3faeg7epfpWXbgaHrRWaWBKMEqLdy38+svGo=";
   };
+
+  patches = [
+    (fetchpatch {
+      url = "https://sources.debian.org/data/main/f/firewalk/5.0-6/debian/patches/060_fix-ftbfs-gcc-15.patch";
+      hash = "sha256-KEYHTSnUtNIGqm/uE2ZLV79KIWmofHeKVYdfTkpm4rI=";
+    })
+  ];
 
   buildInputs = [
     libnet

